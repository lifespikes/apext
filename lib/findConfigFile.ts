import process from 'process'
import colors from '@colors/colors/safe'
import { join } from 'path'
import { isDirectory } from './isDirectory'
import fs from 'fs'

export const findConfigFile = (): any => {
  const routePath = join(process.cwd())
  if (isDirectory(routePath)) {
    // check if file exists
    const jsConfig = `${routePath}/apext.config.js`
    const tsConfig = `${routePath}/apext.config.ts`
    const jsPath = fs.existsSync(jsConfig)
    const tsPath = fs.existsSync(tsConfig)

    if (!jsPath && !tsPath) {
      return {}
    }

    if (jsPath) {
      return require(jsConfig)
    }

    if (tsPath) {
      return require(tsConfig)
    }
    return {}
  }

  throw colors.red(
    "ERROR: Please run this in your Next.js project root directory. \nIf you are in your Next.js project, we couldn't find the 'pages' directory.\n"
  )
}
