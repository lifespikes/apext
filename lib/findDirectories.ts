import { resolve } from 'path'
import { dirExists } from './dirExists'
import colors from '@colors/colors/safe'

export const findDirectories = (
  pathOption: string = ''
): {
  rootPath: string
  pagesPath: string
  apiPath: string
} => {
  const isTest = pathOption.includes('test')
  let rootPath
  let pagesPath
  if (pathOption && !isTest) {
    const specificPath = resolve('src', 'pages', 'api', pathOption)
    if (!dirExists(specificPath)) {
      throw colors.red(
        `The path ${pathOption} does not exist. Please specify a valid path under the /api directory.`
      )
    }

    return {
      apiPath: specificPath,
      pagesPath: resolve('src', 'pages'),
      rootPath: resolve('src')
    }
  }

  rootPath = resolve(pathOption)
  console.log(rootPath)
  pagesPath = resolve(pathOption, 'pages')
  if (!dirExists(pagesPath)) {
    pagesPath = resolve(pathOption, 'src', 'pages')
    rootPath = resolve(rootPath, 'src')
  }
  if (!dirExists(pagesPath)) {
    throw colors.red(
      "ERROR: Please run this in your Next.js project root directory or provide a valid path. \nIf you are in your Next.js project, we couldn't find the 'pages' directory.\n"
    )
  }
  const apiPath = resolve(pagesPath, 'api')
  if (!dirExists(apiPath)) {
    throw colors.red(
      "ERROR: We could not find the 'api' directory in your 'pages' directory. \n You either don't have one or you are running this CLI outside your Next.js root directory.\n"
    )
  }
  return {
    rootPath,
    pagesPath,
    apiPath
  }
}
