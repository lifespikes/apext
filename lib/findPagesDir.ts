import process from 'process'
import colors from '@colors/colors/safe'
import { join } from 'path'
import { isDirectory } from './isDirectory'

export const findPagesDir = async (path: string = ''): Promise<string> => {
  const routePath = join(process.cwd(), path)
  const paths = [join(routePath, 'pages'), join(routePath, 'src', 'pages')]
  for (const path of paths) {
    try {
      if (isDirectory(path)) {
        return path
      }
    } catch (error: any) {}
  }
  throw colors.red(
    "ERROR: Please run this in your Next.js project root directory. \nIf you are in your Next.js project, we couldn't find the 'pages' directory.\n"
  )
}
