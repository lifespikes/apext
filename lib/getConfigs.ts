import process from 'process'
import colors from '@colors/colors/safe'
import { join } from 'path'
import { isDirectory } from './isDirectory'
import fs from 'fs'
import { Configs } from '../types'
import defaultConfig from '../defaultConfig'

export const getConfigs = (): Configs => {
  const routePath = join(process.cwd())
  if (isDirectory(routePath)) {
    // check if file exists
    const jsConfig = `${routePath}/apext.config.js`
    const tsConfig = `${routePath}/apext.config.ts`
    const jsPath = fs.existsSync(jsConfig)
    const tsPath = fs.existsSync(tsConfig)

    if (jsPath) {
      return { ...defaultConfig, ...require(jsConfig) }
    }

    if (tsPath) {
      return { ...defaultConfig, ...require(tsConfig) }
    }
    return { ...defaultConfig }
  }

  throw colors.red(
    "ERROR: Please run this in your Next.js project root directory. \nIf you are in your Next.js project, we couldn't find the 'pages' directory.\n"
  )
}
