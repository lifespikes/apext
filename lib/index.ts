export * from './findApiDir'
export * from './findPagesDir'
export * from './traverseDir'
export * from './isDirectory'
export * from './checkFileExists'
export * from './createFile'
export * from './createFolder'
export * from './findDirectories'
export * from './findConfigFile'
