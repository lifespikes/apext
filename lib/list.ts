import { Options } from '../types'
import { findDirectories } from './findDirectories'
import { traverseDir } from './traverseDir'

export const list = async (options: Options): Promise<string[]> => {
  const directories = findDirectories(options.path)
  return await traverseDir(directories.rootPath, directories.apiPath)
}
