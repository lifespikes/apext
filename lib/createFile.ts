import colors from '@colors/colors/safe'
import { writeFile } from 'fs/promises'
import { toCamelCase } from './sanitizeRoutes'
import { getConfigs } from './getConfigs'

export const createFile = async (
  filePath: string,
  language: string
): Promise<void> => {
  // CONFIG OPTIONS
  const { jsContent, tsContent, typescript } = getConfigs()

  const file = filePath.split('/').pop() as string // filename.extension
  const filename = toCamelCase(file.split('.')[0] ?? '') // filename
  if (typescript) {
    await writeFile(filePath, tsContent(filename))
  } else {
    if (language === 'ts') {
      await writeFile(filePath, tsContent(filename))
    }
    if (language === 'js') {
      await writeFile(filePath, jsContent(filename))
    }
  }

  console.log(' ')
  console.log(colors.green('Route created successfully!'))
  console.log(
    colors.blue('Click to open the route') +
      colors.yellow(' => ') +
      filePath.split('/pages')[1]
  )
  console.log(' ')
}
