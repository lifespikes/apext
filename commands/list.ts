import { Options } from '../types'
import colors from '@colors/colors/safe'
import { findApiDir, findPagesDir, traverseDir } from '../lib'

export const listAction = async (options: Options): Promise<void> => {
  try {
    if (options.path) {
      // output all API routes in the given folder
      const pagesPath = await findPagesDir()
      const apiPath = await findApiDir(pagesPath)
      const path: string = options.path
      const pathWithFolder = `${apiPath}/${path}`
      await traverseDir(pathWithFolder, `/api/${path}`)
    } else {
      // output all API routes in the project
      const pagesPath = await findPagesDir()
      const apiPath = await findApiDir(pagesPath)
      await traverseDir(apiPath)
    }
  } catch (error) {
    console.log(error)
  }
}

export const listDesc = `ðŸŽ¯ ${colors.green(
  'List all API endpoints in your Next.js project.'
)}`
