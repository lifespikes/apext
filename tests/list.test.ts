import { it, expect } from 'vitest'
import { list } from '../lib/list'
import { replaceArrows } from './replaceArrows'

it('lists files from root level pages', async () => {
  const res = (await list({ path: 'tests/fixtures/root-level-pages' })).map(replaceArrows)
  expect(res).toMatchSnapshot()
})

it('lists files from src level pages', async () => {
  const res = (await list({ path: 'tests/fixtures/src-level-pages' })).map(replaceArrows)
  expect(res).toMatchSnapshot()
})
