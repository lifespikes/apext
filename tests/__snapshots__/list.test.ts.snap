// Vitest Snapshot v1

exports[`lists files from root level pages 1`] = `
[
  "/api/[32mfile-no-comments[39m -> [35mSee: https://apext.vercel.app/docs/methods[39m",
  "/api/[32mfile1[39m -> [34mGET[39m|[33mPOST[39m|[33mPUT[39m|[31mDELETE[39m",
  "/api/nested/[32ma[39m -> [34mGET[39m|[33mPOST[39m|[33mPUT[39m|[31mDELETE[39m",
  "/api/nested/[32mb[39m -> [34mGET[39m|[33mPOST[39m|[33mPUT[39m|[31mDELETE[39m",
]
`;

exports[`lists files from src level pages 1`] = `
[
  "/api/[32mfile-no-comments[39m -> [35mSee: https://apext.vercel.app/docs/methods[39m",
  "/api/[32mfile1[39m -> [34mGET[39m|[33mPOST[39m|[33mPUT[39m|[31mDELETE[39m",
  "/api/nested/[32ma[39m -> [34mGET[39m|[33mPOST[39m|[33mPUT[39m|[31mDELETE[39m",
  "/api/nested/[32mb[39m -> [34mGET[39m|[33mPOST[39m|[33mPUT[39m|[31mDELETE[39m",
]
`;
